packages/lodestar/src/api/impl/beacon/state/utils.ts

```diff
@@ -198,3 +204,46 @@ export function filterStateValidatorsByStatuses(
   }
   return responses;
 }
+
+/**
+ * Get the archived state nearest to `slot`.
+ */
+function getNearestArchivedState(
+  config: IBeaconConfig,
+  stateCache: StateContextCache,
+  forkChoice: IForkChoice,
+  slot: Slot
+): CachedBeaconState<allForks.BeaconState> {
+  const nearestArchivedStateSlot = slot - (slot % (PERSIST_STATE_EVERY_EPOCHS * config.params.SLOTS_PER_EPOCH));
+
+  let state: CachedBeaconState<allForks.BeaconState> | null = null;
+  const blockSummary = forkChoice.getCanonicalBlockSummaryAtSlot(nearestArchivedStateSlot);
+  if (blockSummary) {
+    state = stateCache.get(blockSummary.stateRoot);
+  }
+  if (state === null) throw new Error("getNearestArchivedState: cannot find state");
+  return state;
+}
+
+async function getFinalizedState(
+  config: IBeaconConfig,
+  db: IBeaconDb,
+  stateCache: StateContextCache,
+  forkChoice: IForkChoice,
+  slot: Slot
+): Promise<CachedBeaconState<allForks.BeaconState>> {
+  assert.lte(slot, forkChoice.getFinalizedCheckpoint().epoch * config.params.SLOTS_PER_EPOCH);
+  let state = getNearestArchivedState(config, stateCache, forkChoice, slot);
+
+  // process blocks up to the requested slot
+  for await (const block of db.blockArchive.valuesStream({gt: state.slot, lte: slot})) {
+    state = fast.fastStateTransition(state, block);
+    // yield to the event loop
+    await sleep(0);
+  }
```

Comment:
@wemeetagain Should we run this through the regen to queue things?

Reply:
Also this can destroy the node by overloading it. This functionality should not be exposed by default.

Reply:
We don't want to use regen bc regen (and `runStateTransition`) emits events meant for the handling of unfinalized blocks. These blocks and state are completely useless to the function of the chain since its already been finalized, so we want to run this more ephemerally.
This definitely can wreck our node and would only be available behind the `debug` api.

Reply:
i changed this PR to where this functionality is only called when using debug api